src/Microsoft.AspNetCore.Razor.Evolution/TagHelperBinderSyntaxTreePass.cs

```diff
@@ -4,13 +4,17 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
-using System.Threading.Tasks;
 using Microsoft.AspNetCore.Razor.Evolution.Legacy;
 
 namespace Microsoft.AspNetCore.Razor.Evolution
 {
     internal class TagHelperBinderSyntaxTreePass : IRazorSyntaxTreePass
     {
+        private static HashSet<char> InvalidNonWhitespaceNameCharacters = new HashSet<char>(new[]
+        {
+            '@', '!', '<', '/', '?', '[', '>', ']', '=', '"', '\'', '*'
+        });
```

Comment:
This was copied from the workspaces code.

When we define the new THD abstraction, it should be part of that.